zipkin-server/src/main/java/zipkin2/server/internal/elasticsearch/LazyHttpClientImpl.java

```diff
@@ -0,0 +1,158 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin2.server.internal.elasticsearch;
+
+import com.linecorp.armeria.client.ClientFactory;
+import com.linecorp.armeria.client.ClientOptionsBuilder;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.HttpClient;
+import com.linecorp.armeria.client.HttpClientBuilder;
+import com.linecorp.armeria.client.encoding.HttpDecodingClient;
+import com.linecorp.armeria.client.endpoint.EndpointGroup;
+import com.linecorp.armeria.client.endpoint.EndpointGroupRegistry;
+import com.linecorp.armeria.client.endpoint.StaticEndpointGroup;
+import com.linecorp.armeria.client.endpoint.healthcheck.HttpHealthCheckedEndpointGroup;
+import com.linecorp.armeria.client.endpoint.healthcheck.HttpHealthCheckedEndpointGroupBuilder;
+import com.linecorp.armeria.client.logging.LoggingClientBuilder;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.logging.LogLevel;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import zipkin2.elasticsearch.ElasticsearchStorage.LazyHttpClient;
+import zipkin2.server.internal.elasticsearch.ZipkinElasticsearchStorageProperties.HttpLogging;
+
+import static com.linecorp.armeria.client.endpoint.EndpointSelectionStrategy.ROUND_ROBIN;
+
+final class LazyHttpClientImpl implements LazyHttpClient {
+  final SessionProtocol protocol;
+  final Supplier<EndpointGroup> initialEndpoints;
+  /**
+   * Customizes the {@link HttpClientBuilder} used when connecting to ElasticSearch. This is used by
+   * the server and tests to enable detailed logging and tweak timeouts. It is also used by
+   * zipkin-aws to customize authentication of requests.
+   */
+  final List<Consumer<ClientOptionsBuilder>> customizers;
+  final ClientFactory factory;
+  final int timeout;
+  final HttpLogging httpLogging;
+
+  volatile HttpClient result;
+
+  LazyHttpClientImpl(ClientFactory factory, SessionProtocol protocol,
+    Supplier<EndpointGroup> initialEndpoints, List<Consumer<ClientOptionsBuilder>> customizers,
+    int timeout, HttpLogging httpLogging) {
+    this.protocol = protocol;
+    this.initialEndpoints = initialEndpoints;
+    this.customizers = customizers;
+    this.factory = factory;
+    this.timeout = timeout;
+    this.httpLogging = httpLogging;
+  }
+
+  @Override public void close() {
+    EndpointGroup endpointGroup = EndpointGroupRegistry.get("elasticsearch");
+    if (endpointGroup != null) {
+      endpointGroup.close();
+      EndpointGroupRegistry.unregister("elasticsearch");
+    }
+    factory.close();
```

Comment:
Could put this in a `finally` or local try-with-resources in case `endpointGroup.close()` throws for some reason, though it's extremely careful.

Reply:
factory is now external and closed via spring on shutdown. good catch though.